### PR TITLE
Support linux, add cmake and docker support

### PR DESCRIPTION
**Pull request by marcschier**
* Add cmake build system for linux and windows. Removed VS 2013 sln. Update
      readme.md and use latest ansi-c stack from master.
    * Windows build uses dnssd.dll from dnssd repo rather than checked in Apple
      code.
    * Update to only poll in 2 places (todo: 1 poll loop with multiple fd) and
      synchronized access to sdref between poll thread (main) and server threads
      to fix race conditions and crashes. Removed obsolete reentrant locking.
    * Add platform support for Linux using Avahi dnssd adapter library.
    * Add dockerfile to create base image with lds-me support (for .net stack).
    * Simplfied debugging and trouble shooting:
    	* Default to -d when _DEBUG defined
    	* Precreate file paths for pki if they do not exist. Fallback to use cwd
      	  if configured file path for pki does not exist and cannot be created.
    	* Set tracing in config files for stack to "error", not off.
    * Make lds static registration permanent without knowing the semfile by
      allowing to specify * as semfile. If semfile does not exist lds entry is
      immediately removed.

### CODE REVIEW

File: findserversonnetwork.c [hunk 1]
```diff
@@ -171,71 +149,61 @@ void ualds_findserversonnetwork_removeServiceEntries(ualds_browseContext *pBrows
...
+    OpcUa_Mutex_Lock(g_mutex);
     OpcUa_List_Enter(&g_findServersSocketList);
     OpcUa_List_ResetCurrent(&g_findServersSocketList);
 
-    MulticastSocketCallbackStruct* socketCallbackStruct = OpcUa_Null;
-
     socketCallbackStruct = (MulticastSocketCallbackStruct*)OpcUa_List_GetCurrentElement(&g_findServersSocketList);
     while (socketCallbackStruct)
     {
-        if (*shutdown)
```
Comment (hansgschossmann) [inline findserversonnetwork.c]: Need to check if shutdown == null?

File: findserversonnetwork.c [hunk 2]
```diff
@@ -171,71 +149,61 @@ void ualds_findserversonnetwork_removeServiceEntries(ualds_browseContext *pBrows
...
-        fd_set         readFDs;
-        struct timeval tv;
-
+        fd = DNSServiceRefSockFD(socketCallbackStruct->sdRef);
         FD_ZERO(&readFDs);
-        FD_SET(socketCallbackStruct->fd, &readFDs);
+        FD_SET(fd, &readFDs);
 
         tv.tv_sec = 0;
         tv.tv_usec = 100000;
```
Comment (hansgschossmann) [inline findserversonnetwork.c]: Using a constant would make sense, since it is used a couple of times in the file.

File: findserversonnetwork.c [hunk 3]
```diff
@@ -171,71 +149,61 @@ void ualds_findserversonnetwork_removeServiceEntries(ualds_browseContext *pBrows
...
     OpcUa_List_ResetCurrent(&g_findServersSocketList);
 
-    MulticastSocketCallbackStruct* socketCallbackStruct = OpcUa_Null;
-
     socketCallbackStruct = (MulticastSocketCallbackStruct*)OpcUa_List_GetCurrentElement(&g_findServersSocketList);
     while (socketCallbackStruct)
     {
-        if (*shutdown)
+        context = (ualds_genericContext*)socketCallbackStruct->context;
+        if (*shutdown || context == OpcUa_Null || context->sdRef == OpcUa_Null)
```
Comment (hansgschossmann) [inline findserversonnetwork.c]: Need to check if shutdown != null

**Comment by marcschier** [comment]
Added build scripts for v120_xp based on code review feedback.  Also fixed build break when no avahi is installed.

File: findserversonnetwork.c [hunk 4]
```diff
@@ -240,21 +240,58 @@ void DNSSD_API ualds_DNSServiceResolveReply(DNSServiceRef           sdRef,
...
+    char dotLocaldot[] = ".local.";
+    int _endsWithDotLocalDot = string_ends_with(resolvedHostName, dotLocaldot);
+    if (_endsWithDotLocalDot == 0)
     {
-        OpcUa_String_StrnCat(&pRecord->DiscoveryUrl,
-            OpcUa_String_FromCString((OpcUa_StringA)hosttarget),
-            strlen(hosttarget) - strlen(dot));
+        // hosttarget == LABELS.local.
+        memset(resolvedHostName, 0, UALDS_CONF_MAX_URI_LENGTH);
+        strncpy(resolvedHostName, hosttarget, strlen(hosttarget) - strlen(dotLocaldot));
```
Comment (bernd-edlinger) [inline findserversonnetwork.c]: Please be extra careful when subtracting strlen(s)

**Pull request merged**